File: Intermediate/8.Implement binary search for an unsorted array/Solution.c
#include <stdio.h>
#include <stdlib.h>

// Function to sort the array (Bubble Sort)
void sortArray(int arr[], int n) {
    for (int i = 0; i < n - 1; i++) {
        for (int j = 0; j < n - i - 1; j++) {
            if (arr[j] > arr[j + 1]) {
                int temp = arr[j];
                arr[j] = arr[j + 1];
                arr[j + 1] = temp;
            }
        }
    }
}

// Binary Search function
int binarySearch(int arr[], int n, int target) {
    int left = 0, right = n - 1;

    while (left <= right) {
        int mid = left + (right - left) / 2;

        if (arr[mid] == target) return mid;
        else if (arr[mid] < target) left = mid + 1;
        else right = mid - 1;
    }

    return -1;
}

int main() {
    int n, target;

    printf("Enter the number of elements: ");
    scanf("%d", &n);

    int arr[n];
    printf("Enter the unsorted elements:\n");
    for (int i = 0; i < n; i++) {
        scanf("%d", &arr[i]);
    }

    printf("Enter the target element: ");
    scanf("%d", &target);

    sortArray(arr, n); // Sort the array

    int result = binarySearch(arr, n, target);

    if (result != -1) {
        printf("Element found at index %d (after sorting).\n", result);
    } else {
        printf("Element not found.\n");
    }

    return 0;
}

File: Intermediate/8.Implement binary search for an unsorted array/Solution.java
import java.util.Arrays;
import java.util.Scanner;

public class BinarySearchUnsorted {
    static int binarySearch(int[] arr, int target) {
        int left = 0, right = arr.length - 1;

        while (left <= right) {
            int mid = left + (right - left) / 2;

            if (arr[mid] == target) return mid;
            else if (arr[mid] < target) left = mid + 1;
            else right = mid - 1;
        }

        return -1; // Target not found
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.print("Enter the number of elements: ");
        int n = scanner.nextInt();

        int[] arr = new int[n];
        System.out.println("Enter the unsorted elements:");
        for (int i = 0; i < n; i++) {
            arr[i] = scanner.nextInt();
        }

        System.out.print("Enter the target element: ");
        int target = scanner.nextInt();

        Arrays.sort(arr); // Sort the array

        int result = binarySearch(arr, target);

        if (result != -1) {
            System.out.println("Element found at index " + result + " (after sorting).");
        } else {
            System.out.println("Element not found.");
        }

        scanner.close();
    }
}
